### ISSUE TITLE
Owen Q-function in scipy

### ISSUE THREAD

**Issue by stla** [scipy/scipy#15116]
### Is your feature request related to a problem? Please describe.

Hello,

  I have implemented the [Owen Q-function](https://gist.github.com/stla/55d9ff60ab9b06eafedf403734b31684) in Python, using `scipy`. Are you interested in a PR? This function is important in statistics because it allows to calculate the power of an equivalence test (see the [vignettes here](https://cran.r-project.org/web/packages/OwenQ/index.html)).

**Comment by tupui** [comment]
Hi @stla, thanks for the proposal. Before going further. I see that you are the author of the R package you linked. For inclusion in SciPy, you would need to re-license explicitly your work to a [BSD 3 clauses](https://github.com/scipy/scipy/blob/master/LICENSE.txt). Is that ok?

**Comment by stla** [comment]
Ok, will do.

I also have some [code](https://gist.github.com/stla/6e2cbd868825d169d0db0b2c99a5e704) to sample the noncentral Wishart distribution. It is taken from my R package ***matrixsampling***. A particular feature is the ability to sample the noncentral Wishart when the number of degrees of freedom is lower than 2p-1 (p is the dimension). I have been able to implement this feature thanks to a recent paper.

**Comment by stla** [comment]
Hello @tupui ,

  I have relicensed the package **OwenQ** to BSD 3.

**Comment by tupui** [comment]
Could you describe in practical terms how this would be used? Reading at what you linked, I am not sure about what it does.

**Comment by stla** [comment]
Ok, but later. If you want, you can take a look at the vignettes (I posted the links). One of them give the formula to get the power of the TOST, which is the target application of this implementation.

**Comment by josef-pkt** [comment]
related open issue for statsmodels, exact power for t-TOST equivalence testing
https://github.com/statsmodels/statsmodels/issues/6822

**Comment by stla** [comment]
Well @tupui to sum up, the power of the equivalence test is computed from the Owen-`O4` function, denoted by `powen4` in my R package:

![OwenQ_powen4](https://user-images.githubusercontent.com/4466543/148544089-e495c433-0e14-4ff1-a0fd-6bc4588ad79a.png)

Then Owen-`Q1` function, which is the function implemented in my Python code is defined as follows:

![OwenQfunction](https://user-images.githubusercontent.com/4466543/148544253-f8e55ef1-7fa4-4516-969e-846435691c3a.png)

And one has the following equality:

![OwenQ_O4](https://user-images.githubusercontent.com/4466543/148544320-7fb4e7a2-e2c8-4ab8-8872-89510f69cf5b.png)

Is it clearer now?

**Comment by tupui** [comment]
Ok so if I try to rephrase: it gives the power, probability of true positive, to correctly reject the null hypothesis. Which in this case goes to ask if the two samples have the same variance. Is that right?

**Comment by stla** [comment]
No, this has nothing to do with the variances. The goal of the test is to prove (accept H1) that the difference between the means of the two samples is not too large (for a threshold given by the user).

**Comment by stla** [comment]
Yes, there is the normal assumption.

The idea of the TOST (two one-sided tests), an *equivalence test*, as opposed as a *difference test*, is that it is able to prove (up to a probability error, of course) that the difference between two means fall into a given range. Whereas a difference test proves that two means are different, that is, their difference is not zero. The problem with such a test is that you will always find a difference different from zero if you increase the sample sizes.

**Comment by tupui** [comment]
Thanks for the details. This looks useful, 👍 on my side. @mdhaber what do you think?

We will need to work on the API and documentation (with a real life example) to make it accessible. As you've seen, it took a bit of time for someone like me to get it 😉

**Comment by stla** [comment]
Maybe there are some real-life examples in the vignettes of the R package [PowerTOST](https://cran.r-project.org/web/packages/PowerTOST/index.html).

**Comment by josef-pkt** [comment]
this issue is for adding another `special` function.

statsmodels has TOST equivalence tests versions of many of the hypothesis tests already, and power for them, but only the standard, and  not the "exact" method that R has.

**Comment by tupui** [comment]
I am not sure about what you want to say @josef-pkt. Do you mean that are -1 for inclusion here and that's more on the scope of statsmodels?

**Comment by josef-pkt** [comment]
No, what I'm saying is:

It would be very useful to have Owens Q as new scipy.special function.
But it is not necessary that scipy.stats goes into TOST functions, because that's largely covered by statsmodels.
Or, if scipy.stats will get TOST, then it's a decision independent of adding Owens q.

**Comment by mdhaber** [comment]
Sounds like `scipy.special` is the way to go.

**Comment by mdhaber** [comment]
@josef-pkt Can you suggest ranges of parameters with which this function might be called if you were to use it in statsmodels? If so, I can put something together and test it to make sure it's robust enough for your needs.

---

The integrand looks very well-behaved to me, and it is probably fine to just integrate the function numerically.

For example, here is an example calculation with the R `OwenQ` package:

<details>

```R
library(OwenQ)
options(digits=16)
OwenQ1(nu=10, t=3, delta=2, R=1)
# 2.190096901078659e-05
```

</details>

and a simple Python implementation:

<details>

```python3
import numpy as np
from scipy import special, integrate
from scipy.integrate._tanhsinh import _tanhsinh
from scipy.special import ndtr as G, log_ndtr as logG

sqrt_2_pi = np.sqrt(2 * np.pi)

def Gprime(x):
    return 1/sqrt_2_pi * np.exp(-0.5 * x**2)

def q(x, f, t, d):
    return G(t*x/np.sqrt(f) - d) * x**(f-1) * Gprime(x)

def Q(f, t, d, a, b):
    den = special.gamma(0.5 * f) * 2**(0.5*(f - 2))
    C = sqrt_2_pi / den
    res = _tanhsinh(q, a, b, args=(f, t, d))
    return res.integral * C
```

</details>

```python3
# With the t, d, and r used in OwenQ vignette
f, t, d, R = 10, 3, 2, 1
res = Q(f, t, d, 0, R)
ref = 2.190096901078659e-05  # reference from R OwenQ
(res - ref)/ref
# 1.8765397357777725e-13
```

This would be somewhat slow by `special` function standards (~1ms for one function evaluation, ~3ms for 100 in a vectorized call), but perhaps good enough for the sorts of things it would be used for.
<details>

```python3
# %timeit Q(f, t, d, 0, R)
# 1.17 ms ± 94.4 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

# But note that Q above is vectorized:
# R = np.linspace(-5, 5, 100)
# %timeit Q(f, t, d, 0, R)
# 3.14 ms ± 694 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

</details>

The `1e-13` relative error surprised me. However, I think it's actually the Python implementation that's more accurate: if we trust `mpmath` to do the integral (`mp.dps = 50`).

<details>

```python3
from mpmath import mp
mp.dps = 50

sqrt_2_pi_mp = mp.sqrt(2 * mp.pi)

def q_mp(x, f, t, d):
    return mp.ncdf(t*x/mp.sqrt(f) - d) * x**(f - mp.one) * mp.npdf(x)

def Q_mp(f, t, d, a, b):
    f, t, d = mp.mpf(f), mp.mpf(t), mp.mpf(d)
    C = sqrt_2_pi_mp / (mp.gamma(f/2) * 2**((f - 2)/2))
    res = mp.quad(lambda x: q_mp(x, f, t, d), (a, b))
    return float(res * C)
```

</details>

```python3
ref_mp = Q_mp(f, t, d, 0, R)
(res - ref_mp)/ref_mp
# 1.5470236898412373e-16
```

So I think we're actually getting close to machine precision in this example.

Also, it looks like the recurrence relation in the paper has a few disadvantages:
- Time complexity looks at least proportional to `f`
- It might not be numerically favorable for small probabilities
- It doesn't work for fractional `f` (not sure if that matters at all in applications)

I mention the numerical properties assuming `OwenQ` uses the recurrence and that the accuracy is representative of the algorithm itself, although I don't know if either of those things are true. I just noticed in limited testing accuracy for small probabilities seems to be poor compared to the mpmath implementation (e.g. just increase `f` in the tests above).

Note that in case the log of the function would be useful (for very large `f` or extremely small probabilities), we can do the whole calculation in log-space.

<details>

```python3
log_sqrt_2_pi = np.log(sqrt_2_pi)

def logGprime(x):
    return -log_sqrt_2_pi - 0.5*x**2

def logq(x, f, t, d):
    return logG(t*x/np.sqrt(f) - d) + (f-1)*np.log(x) + logGprime(x)

def logQ(f, t, d, a, b):
    den = special.gammaln(0.5 * f) + (0.5 * (f-2))*np.log(2)
    C = log_sqrt_2_pi - den
    res = _tanhsinh(logq, a, b, args=(f, t, d), log=True)
    return res.integral + C

res = logQ(f, t, d, 0, R)
ref = float(mp.log(Q_mp(f, t, d, 0, R)))
(res - ref)/ref
# 1.6556624145077783e-16
```

</details>

**Comment by wtwang1998** [comment]
Is there any progress on this issue? I see that there is already an `owens_t` function in `scipy.special`.
